flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TimeBoundedStreamInnerJoin.scala

```diff
@@ -0,0 +1,533 @@
+        }
+
+        // Join the record with records from the opposite stream.
+        val oppositeIterator = oppositeCache.iterator()
+        var oppositeEntry: Entry[Long, util.List[Row]] = null
+        var oppositeTime: Long = 0L;
+        while (oppositeIterator.hasNext) {
+          oppositeEntry = oppositeIterator.next
+          oppositeTime = oppositeEntry.getKey
+          if (oppositeTime < oppositeLowerBound - allowedLateness) {
+            //TODO Considering the data out-of-order, we should not remove records here.
+          } else if (oppositeTime >= oppositeLowerBound && oppositeTime <= oppositeUpperBound) {
+            val oppositeRows = oppositeEntry.getValue
+            var i = 0
+            if (leftRecord) {
+              while (i < oppositeRows.size) {
+                joinFunction.join(record, oppositeRows.get(i), cRowWrapper)
+                i += 1
+              }
+            } else {
+              while (i < oppositeRows.size) {
+                joinFunction.join(oppositeRows.get(i), record, cRowWrapper)
+                i += 1
+              }
+            }
+          } else if (oppositeTime > oppositeUpperBound) {
+            //TODO If the keys are ordered, can we break here?
+          }
+        }
+      } else {
+        //TODO Need some extra logic here?
+        LOG.warn(s"$record is out-of-date.")
+      }
+    }
+  }
+
+  /**
+    * Register a timer for cleaning up records in a specified time.
+    *
+    * @param ctx               the context to register timer
+    * @param timeForRecord     time for the input record
+    * @param oppositeWatermark watermark of the opposite stream
+    * @param timerState        stores the timestamp for the next timer
+    * @param leftRecord        record from the left or the right stream
+    * @param firstTimer        whether this is the first timer
+    */
+  private def registerCleanUpTimer(
+    ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+    timeForRecord: Long,
+    oppositeWatermark: Long,
+    timerState: ValueState[Long],
+    leftRecord: Boolean,
+    firstTimer: Boolean): Unit = {
+    val cleanUpTime = timeForRecord + (if (leftRecord) leftRelativeSize else rightRelativeSize) +
+      allowedLateness + 1
+    registerTimer(ctx, !leftRecord, cleanUpTime)
+    LOG.debug(s"Register a clean up timer on the ${if (leftRecord) "RIGHT" else "LEFT"} state:"
+      + s" timeForRecord = ${timeForRecord}, cleanUpTime = ${cleanUpTime}, oppositeWatermark = " +
+      s"${oppositeWatermark}")
+    timerState.update(cleanUpTime)
+    if (cleanUpTime <= oppositeWatermark + allowedLateness && firstTimer) {
+      backPressureSuggestion =
+        if (leftRecord) (oppositeWatermark + allowedLateness - cleanUpTime)
+        else -(oppositeWatermark + allowedLateness - cleanUpTime)
+      LOG.warn("The clean timer for the " +
+        s"${if (leftRecord) "left" else "right"}" +
+        s" stream is lower than ${if (leftRecord) "right" else "left"} watermark." +
```

Comment:
its not a problem if the clean up timer is lower than the watermark. In that case, the timer fires immediately.

I'm not sure about the purpose of the backpressure suggestion. AFAIK, there is no mechanism to control the backpressure in Flink. It's rather a self-regulating mechanism.